### ISSUE TITLE
All subdomains go to same docker instance

### ISSUE THREAD

**Issue by sjensen85** [dokku/dokku#2902]
I am not sure this is an issue, but could very well be a limitation or a misconfiguration.

Description of problem:
I have successfully deployed many apps to dokku. One being a React app with create-react-app and mars's create-react-app buildpack and gitlab-ce with a Docker image. The problem I am facing is that when either of these is deployed, all subdomains go to the same app.

For instance with Gitlab:
I go to gitlab.subdomain.domain.com and I get Gitlab.
I go to test.subdomain.domain.com (where the React app should be reachable when deployed) and I get Gitlab.
Also going to subdomain.domain.com goes to Gitlab.

This happens with both apps deployed or just one of the apps. The React app behaves the same way when deployed (with or without Gitlab deployed). It depends on which app gets deployed first, whichever it is, it will take over all sub-subdomains.

I've looked at the various nginx.confs and they all seem to be configured properly.

I have also tried only using it as a heroku style PaaS and deploying multiple React apps, same issue.

Output of dokku report:

> -----> uname: Linux domain.com 3.19.0-65-generic #73~14.04.1-Ubuntu SMP Wed Jun 29 21:05:22 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux
> -----> memory:
>              total       used       free     shared    buffers     cached
> Mem:         14024      11533       2490         32        963       5932
> -/+ buffers/cache:       4638       9386
> Swap:            0          0          0
> -----> docker version:
> Client:
>  Version:      17.07.0-ce
>  API version:  1.31
>  Go version:   go1.8.3
>  Git commit:   8784753
>  Built:        Tue Aug 29 17:43:06 2017
>  OS/Arch:      linux/amd64
> 
> Server:
>  Version:      17.07.0-ce
>  API version:  1.31 (minimum version 1.12)
>  Go version:   go1.8.3
>  Git commit:   8784753
>  Built:        Tue Aug 29 17:41:57 2017
>  OS/Arch:      linux/amd64
>  Experimental: false
> -----> docker daemon info:
> Containers: 57
>  Running: 1
>  Paused: 0
>  Stopped: 56
> Images: 93
> Server Version: 17.07.0-ce
> Storage Driver: aufs
>  Root Dir: /var/lib/docker/aufs
>  Backing Filesystem: extfs
>  Dirs: 270
>  Dirperm1 Supported: true
> Logging Driver: json-file
> Cgroup Driver: cgroupfs
> Plugins:
>  Volume: local
>  Network: bridge host macvlan null overlay
>  Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
> Swarm: inactive
> Runtimes: runc
> Default Runtime: runc
> Init Binary: docker-init
> containerd version: 3addd840653146c90a254301d6c3a663c7fd6429
> runc version: 2d41c047c83e09a6d61d464906feb2a2f3c52aa4
> init version: 949e6fa
> Security Options:
>  apparmor
> Kernel Version: 3.19.0-65-generic
> Operating System: Ubuntu 14.04.5 LTS
> OSType: linux
> Architecture: x86_64
> CPUs: 2
> Total Memory: 13.7GiB
> Name: subdomain.domain.com
> ID: OCWT:XRLO:IFCF:6UNT:6VCG:2KP2:MPFJ:ESWN:SCLT:4MCD:UUZW:Q3U6
> Docker Root Dir: /var/lib/docker
> Debug Mode (client): true
> Debug Mode (server): false
> Registry: https://index.docker.io/v1/
> Experimental: false
> Insecure Registries:
>  127.0.0.0/8
> Live Restore Enabled: false
> 
> WARNING: No swap limit support
> -----> sigil version: 0.4.0
> -----> herokuish version:
> herokuish: 0.3.32
> buildpacks:
>   heroku-buildpack-multi     v1.0.0
>   heroku-buildpack-ruby      v168
>   heroku-buildpack-nodejs    v99
>   heroku-buildpack-clojure   v77
>   heroku-buildpack-python    v99
>   heroku-buildpack-java      v53
>   heroku-buildpack-gradle    v23
>   heroku-buildpack-scala     v78
>   heroku-buildpack-play      v26
>   heroku-buildpack-php       v122
>   heroku-buildpack-go        v72
>   heroku-buildpack-erlang    fa17af9
>   buildpack-nginx            v8
> -----> dokku version: 0.10.4
> -----> dokku plugins:
> plugn: 0.3.0
>   00_dokku-standard    0.10.4 enabled    dokku core standard plugin
>   20_events            0.10.4 enabled    dokku core events logging plugin
>   apps                 0.10.4 enabled    dokku core apps plugin
>   build-env            0.10.4 enabled    dokku core build-env plugin
>   certs                0.10.4 enabled    dokku core certificate management plugin
>   checks               0.10.4 enabled    dokku core checks plugin
>   common               0.10.4 enabled    dokku core common plugin
>   config               0.10.4 enabled    dokku core config plugin
>   docker-options       0.10.4 enabled    dokku core docker-options plugin
>   domains              0.10.4 enabled    dokku core domains plugin
>   enter                0.10.4 enabled    dokku core enter plugin
>   git                  0.10.4 enabled    dokku core git plugin
>   logs                 0.10.4 enabled    dokku core logs plugin
>   named-containers     0.10.4 enabled    dokku core named containers plugin
>   nginx-vhosts         0.10.4 enabled    dokku core nginx-vhosts plugin
>   plugin               0.10.4 enabled    dokku core plugin plugin
>   proxy                0.10.4 enabled    dokku core proxy plugin
>   ps                   0.10.4 enabled    dokku core ps plugin
>   repo                 0.10.4 enabled    dokku core repo plugin
>   shell                0.10.4 enabled    dokku core shell plugin
>   ssh-keys             0.10.4 enabled    dokku core ssh-keys plugin
>   storage              0.10.4 enabled    dokku core storage plugin
>   tags                 0.10.4 enabled    dokku core tags plugin
>   tar                  0.10.4 enabled    dokku core tar plugin

Gitlab nginx.conf:

```
> server {
>   listen      [::]:22;
>   listen      22;
>   server_name gitlab.subdomain.domain.com;
>   access_log  /var/log/nginx/gitlab-access.log;
>   error_log   /var/log/nginx/gitlab-error.log;
> 
>   location    / {
> 
>     gzip on;
>     gzip_min_length  1100;
>     gzip_buffers  4 32k;
>     gzip_types    text/css text/javascript text/xml text/plain text/x-component application/javascript application/x-javascript application/json application/xml  application/rss+xml font/truetype application/x-font-ttf font/opentype application/vnd.ms-fontobject image/svg+xml;
>     gzip_vary on;
>     gzip_comp_level  6;
> 
>     proxy_pass  http://gitlab-22;
>     proxy_http_version 1.1;
>     proxy_set_header Upgrade $http_upgrade;
>     proxy_set_header Connection "upgrade";
>     proxy_set_header Host $http_host;
>     proxy_set_header X-Forwarded-Proto $scheme;
>     proxy_set_header X-Forwarded-For $remote_addr;
>     proxy_set_header X-Forwarded-Port $server_port;
>     proxy_set_header X-Request-Start $msec;
>   }
>   include /appdata/dokku/gitlab/nginx.conf.d/*.conf;
> 
>   error_page 400 401 402 403 405 406 407 408 409 410 411 412 413 414 415 416 417 418 420 422 423 424 426 428 429 431 444 449 450 451 /400-error.html;
>   location /400-error.html {
>     root /var/lib/dokku/data/nginx-vhosts/dokku-errors;
>     internal;
>   }
> 
>   error_page 404 /404-error.html;
>   location /404-error.html {
>     root /var/lib/dokku/data/nginx-vhosts/dokku-errors;
>     internal;
>   }
> 
>   error_page 500 501 502 503 504 505 506 507 508 509 510 511 /500-error.html;
>   location /500-error.html {
>     root /var/lib/dokku/data/nginx-vhosts/dokku-errors;
>     internal;
>   }
> 
> }
> 
> server {
>   listen      [::]:443;
>   listen      443;
>   server_name gitlab.subdomain.domain.com;
>   access_log  /var/log/nginx/gitlab-access.log;
>   error_log   /var/log/nginx/gitlab-error.log;
> 
>   location    / {
> 
>     gzip on;
>     gzip_min_length  1100;
>     gzip_buffers  4 32k;
>     gzip_types    text/css text/javascript text/xml text/plain text/x-component application/javascript application/x-javascript application/json application/xml  application/rss+xml font/truetype application/x-font-ttf font/opentype application/vnd.ms-fontobject image/svg+xml;
>     gzip_vary on;
>     gzip_comp_level  6;
> 
>     proxy_pass  http://gitlab-443;
>     proxy_http_version 1.1;
>     proxy_set_header Upgrade $http_upgrade;
>     proxy_set_header Connection "upgrade";
>     proxy_set_header Host $http_host;
>     proxy_set_header X-Forwarded-Proto $scheme;
>     proxy_set_header X-Forwarded-For $remote_addr;
>     proxy_set_header X-Forwarded-Port $server_port;
>     proxy_set_header X-Request-Start $msec;
>   }
>   include /appdata/dokku/gitlab/nginx.conf.d/*.conf;
> 
>   error_page 400 401 402 403 405 406 407 408 409 410 411 412 413 414 415 416 417 418 420 422 423 424 426 428 429 431 444 449 450 451 /400-error.html;
>   location /400-error.html {
>     root /var/lib/dokku/data/nginx-vhosts/dokku-errors;
>     internal;
>   }
> 
>   error_page 404 /404-error.html;
>   location /404-error.html {
>     root /var/lib/dokku/data/nginx-vhosts/dokku-errors;
>     internal;
>   }
> 
>   error_page 500 501 502 503 504 505 506 507 508 509 510 511 /500-error.html;
>   location /500-error.html {
>     root /var/lib/dokku/data/nginx-vhosts/dokku-errors;
>     internal;
>   }
> 
> }
> 
> server {
>   listen      [::]:80;
>   listen      80;
>   server_name gitlab.subdomain.domain.com;
>   access_log  /var/log/nginx/gitlab-access.log;
>   error_log   /var/log/nginx/gitlab-error.log;
> 
>   location    / {
> 
>     gzip on;
>     gzip_min_length  1100;
>     gzip_buffers  4 32k;
>     gzip_types    text/css text/javascript text/xml text/plain text/x-component application/javascript application/x-javascript application/json application/xml  application/rss+xml font/truetype application/x-font-ttf font/opentype application/vnd.ms-fontobject image/svg+xml;
>     gzip_vary on;
>     gzip_comp_level  6;
> 
>     proxy_pass  http://gitlab-80;
>     proxy_http_version 1.1;
>     proxy_set_header Upgrade $http_upgrade;
>     proxy_set_header Connection "upgrade";
>     proxy_set_header Host $http_host;
>     proxy_set_header X-Forwarded-Proto $scheme;
>     proxy_set_header X-Forwarded-For $remote_addr;
>     proxy_set_header X-Forwarded-Port $server_port;
>     proxy_set_header X-Request-Start $msec;
>   }
>   include /appdata/dokku/gitlab/nginx.conf.d/*.conf;
> 
>   error_page 400 401 402 403 405 406 407 408 409 410 411 412 413 414 415 416 417 418 420 422 423 424 426 428 429 431 444 449 450 451 /400-error.html;
>   location /400-error.html {
>     root /var/lib/dokku/data/nginx-vhosts/dokku-errors;
>     internal;
>   }
> 
>   error_page 404 /404-error.html;
>   location /404-error.html {
>     root /var/lib/dokku/data/nginx-vhosts/dokku-errors;
>     internal;
>   }
> 
>   error_page 500 501 502 503 504 505 506 507 508 509 510 511 /500-error.html;
>   location /500-error.html {
>     root /var/lib/dokku/data/nginx-vhosts/dokku-errors;
>     internal;
>   }
> 
> }
> 
> upstream gitlab-22 {
> 
>   server 172.17.0.2:22;
> }
> 
> upstream gitlab-443 {
> 
>   server 172.17.0.2:443;
> }
> 
> upstream gitlab-80 {
> 
>   server 172.17.0.2:80;
> }
```

- Link to the exact repository being deployed (if possible/applicable): https://hub.docker.com/r/gitlab/gitlab-ce/

Environment details (AWS, VirtualBox, physical, etc.): Azure

How was Dokku installed?: via apt-get

How reproducible: every app deployed

Steps to Reproduce:
1. Install dokku as instructed in docs.
1. Deploy the the Gitlab image as instructed in docs.
2. Go to expected URL
3. Change sub-subdomain

Actual Results:
Same application with different local stores/sessions/etc.

Expected Results:
Presented with the application that is hosted on the sub-subdomain

Additional info:
As you can tell from the `nginx.conf` file I did change the dokku home directory. This is reflected in `/etc/nginx/conf.d/dokku.conf`. I am also publishing ports 80 for Gitlab as the routing that usually happens when deploying a normal app with buildpack didn't happen. I have played around with the proxy, even turning it off to no avail.

Does dokku work with multiple apps on sub-subdomains? I have not yet tried using paths to configure the apps, that could very well be a good solution, but my team keeps pushing for sub-subdomains. 

I am thinking there is simply a dumb mistake happening, just not sure what. Thanks for any guidance.

**Comment by josegonzalez** [comment]
Dokku definitely works with subdomains.

Are you saying that deploying gitlab after deploying your react app will result in the react app no longer being routed to? Or just that any domain that isn't attached to an app will go to the first app?

**Comment by sjensen85** [comment]
Thanks for the prompt reply. 

If I deploy either app, with no other apps deployed, I can get to the app as expected. So i deploy my react app to test.subdomain.domain.com. I can successfully reach it at that URL. However, I also reach the app by going to testing.subdomain.domain.com or anythinghere.subdomain.domain.com. If I deploy gitlab while the react app is deployed, I cannot get gitlab at all, I still get the react app. I go to gitlab.subdomain.domain.com and I get the React app. Both of these apps are assigned a subdomain - test and gitlab.

I did try the file you mentioned and it broke something so I removed it and moved on (I can't remember the exact error from the time.)

**Comment by sjensen85** [comment]
Maybe this is the issue...

When I first installed dokku, I did it the same way as documented via apt-get. When deploying I kept getting:

>    -----> Creating http nginx.conf
>    -----> Running nginx-pre-reload
>                Reloading nginx
>     remote: reload: Job is not running: nginx

What I had to do was remove dokku via apt-get, make sure nginx could start and reload configurations, then reinstall dokku. 

This fixed that issue for a few days. Now I am always getting this error, even when there is a successful deployment. With gitlab it still deploys and I can reach the app. I believe this is because I am publishing port 80 from the container. However, I this could be why my react app is not reachable.

**Comment by josegonzalez** [comment]
Can you paste the output of the following:

```shell
# replace 'APP' with your app name
dokku config APP | grep DOKKU_
```

**Comment by sjensen85** [comment]
App: test - This is the React app, I cannot reach this at all any longer even with Gitlab stopped:
> DOKKU_APP_RESTORE:    0
> DOKKU_APP_TYPE:       herokuish
> DOKKU_NGINX_PORT:     80
> DOKKU_PROXY_PORT_MAP: http:80:5000

App: gitlab - This does deploy and can reach it
> DOKKU_APP_RESTORE:      0
> DOKKU_DOCKERFILE_PORTS: 22/tcp 443/tcp 80/tcp
> DOKKU_PROXY_PORT_MAP:   http:22:22 http:443:443 http:80:80

**Comment by josegonzalez** [comment]
Do you mind hopping on [slack](https://slack.gliderlabs.com/) so we can figure out whats going on?

**Comment by sjensen85** [comment]
Sure, thanks. I won't be able to until tomorrow though. I will message you.

**Comment by sjensen85** [comment]
@josegonzalez Thanks for all your assistance. I figured out the issue. 

From above, dokku  automatically mapped the port 22 from gitlab to the host port 22. This is why nginx would never complete a reload.

I deleted the gitlab and test app, created new ones, deployed gitlab. Afterwards, I re-mapped port 22 on the host to another port and reloaded the nginx config. I then deployed the same react app and now both are accessible at their respective URLs.

This was an stupid mistake on my end, nothing to do with dokku. Love the product and all the work you all do. Thanks a ton.
